src/ray/gcs/gcs_server/gcs_task_manager.h

```diff
@@ -0,0 +1,161 @@
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "absl/base/thread_annotations.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "absl/synchronization/mutex.h"
+#include "ray/rpc/gcs_server/gcs_rpc_server.h"
+#include "src/ray/protobuf/gcs.pb.h"
+
+namespace ray {
+namespace gcs {
+
+/// Type alias for a single task attempt, i.e. <task id and attempt number>.
+using TaskAttempt = std::pair<TaskID, int32_t>;
+
+/// GcsTaskManger is responsible for capturing task states change reported by
+/// TaskEventBuffer from other components.
+///
+/// When the maximal number of task events tracked specified by
+/// `RAY_task_events_max_num_task_in_gcs` is exceeded, older events (approximately by
+/// insertion order) will be dropped.
+///
+/// This class has its own io_context and io_thread, that's separate from other GCS
+/// services. All handling of all rpc will be posted to the single thread it owns.
+class GcsTaskManager : public rpc::TaskInfoHandler {
+ public:
+  /// Create a GcsTaskManager.
+  GcsTaskManager()
+      : task_event_storage_(std::make_unique<GcsTaskManagerStorage>(
+            RayConfig::instance().task_events_max_num_task_in_gcs())),
+        io_service_thread_(std::make_unique<std::thread>([this] {
+          SetThreadName("task_events");
+          // Keep io_service_ alive.
+          boost::asio::io_service::work io_service_work_(io_service_);
+          io_service_.run();
+        })) {}
+
+  /// Handles a AddTaskEventData request.
+  ///
+  /// \param request gRPC Request.
+  /// \param reply gRPC Reply.
+  /// \param send_reply_callback Callback to invoke when sending reply.
+  void HandleAddTaskEventData(rpc::AddTaskEventDataRequest request,
+                              rpc::AddTaskEventDataReply *reply,
+                              rpc::SendReplyCallback send_reply_callback) override;
+
+  ///  Handles a GetTaskEvents request.
+  ///
+  /// \param request gRPC Request.
+  /// \param reply gRPC Reply.
+  /// \param send_reply_callback Callback to invoke when sending reply.
+  void HandleGetTaskEvents(rpc::GetTaskEventsRequest request,
+                           rpc::GetTaskEventsReply *reply,
+                           rpc::SendReplyCallback send_reply_callback) override {
+    RAY_LOG(FATAL) << "Unimplemented!";
+  };
+
+  /// Stops the event loop and the thread of the task event handler.
+  ///
+  /// This function returns when the io thread is joined.
+  void Stop();
+
+  /// Returns the io_service.
+  ///
+  /// \return Reference to its io_service.
+  instrumented_io_context &GetIoContext() { return io_service_; }
```

Comment:
const instrumented_io_context &GetIoContext() const { return io_service_; }

Reply:
hmm - this needs to be an non-const reference since the [caller](https://github.com/ray-project/ray/blob/561344a084d3e30622b1e2fa560d69755a5ebbf1/src/ray/rpc/grpc_server.h#L151) takes a non-const reference.